Explorer/Assets/DCL/ResourcesUnloading/ECS/ReleaseMemorySystem.cs

```diff
@@ -12,18 +15,44 @@ public partial class ReleaseMemorySystem : BaseUnityLoopSystem
         private readonly IMemoryUsageProvider memoryBudgetProvider;
         private readonly ICacheCleaner cacheCleaner;
 
-        internal ReleaseMemorySystem(Arch.Core.World world, ICacheCleaner cacheCleaner, IMemoryUsageProvider memoryBudgetProvider) : base(world)
+        private readonly IUnloadStrategy[] unloadStrategies;
+        internal int currentUnloadStrategy;
+
+        private int consecutiveFailedFrames;
+        private readonly int failureThreshold;
+
+        internal ReleaseMemorySystem(Arch.Core.World world, ICacheCleaner cacheCleaner,
+            IMemoryUsageProvider memoryBudgetProvider, IUnloadStrategy[] unloadStrategies,
+            int failuresFrameThreshold) : base(world)
         {
             this.cacheCleaner = cacheCleaner;
             this.memoryBudgetProvider = memoryBudgetProvider;
+            this.unloadStrategies = unloadStrategies;
+            failureThreshold = failuresFrameThreshold;
+            currentUnloadStrategy = 0;
         }
 
         protected override void Update(float t)
         {
+            if (unloadStrategies[currentUnloadStrategy].isRunning)
+                return;
+
             if (memoryBudgetProvider.GetMemoryUsageStatus() != MemoryUsageStatus.NORMAL)
-                cacheCleaner.UnloadCache();
+            {
+                unloadStrategies[currentUnloadStrategy].TryUnload(cacheCleaner);
+                consecutiveFailedFrames++;
 
-            cacheCleaner.UpdateProfilingCounters();
+                if (consecutiveFailedFrames >= failureThreshold)
+                {
+                    currentUnloadStrategy = Math.Clamp(currentUnloadStrategy + 1, 0, unloadStrategies.Length - 1);
```

Comment:
maybe even better to have a class for it like `UnloadingStrategy` that would have function `NextTier()` and `ResetToNormal()`. Than it would look even cooler - `unloadingStrategy.NextTier()`, `unloadingStrategy.ResetToNormal()`, `unloadingStrategy.Unload()` and `unloadingStrategy.IsRunnig` (or `.isUnloading`),   but requires additional abstraction, so I am not fully sure :)